File: src/Algorithm/LeetCode/LeetCode_583_Delete_Operation_for_Two_Strings/Solution_DP.java
package Algorithm.LeetCode.LeetCode_583_Delete_Operation_for_Two_Strings;

//https://www.programmercarl.com/0583.两个字符串的删除操作.html
class Solution_DP {
    public int minDistance(String word1, String word2) {
        int length1 = word1.length();
        int length2 = word2.length();

        int[][] dp = new int[length1 + 1][length2 + 1];

        for (int i = 1; i <= length1; i++) {
            for (int j = 1; j <= length2; j++) {
                if (word1.charAt(i - 1) == word2.charAt(j - 1)) {
                    dp[i][j] = dp[i - 1][j - 1] + 1;
                } else {
                    dp[i][j] = Math.max(dp[i - 1][j], dp[i][j - 1]);
                }
            }
        }

        return length1 + length2 - 2 * dp[length1][length2];
    }
}

File: src/Algorithm/LeetCode/LeetCode_115_Distinct_Subsequences/Solution_DP.py
# https://programmercarl.com/0115.不同的子序列.html
class Solution_DP:
    def numDistinct(self, s: str, t: str) -> int:
        length1 = len(s)
        length2 = len(t)

        dp = [[0] * (length2 + 1) for _ in range(length1 + 1)]

        for i in range(length1 + 1):
            dp[i][0] = 1

        for j in range(1, length2 + 1):
            dp[0][j] = 0

        for i in range(1, length1 + 1):
            for j in range(1, length2 + 1):
                if s[i - 1] == t[j - 1]:
                    dp[i][j] = dp[i - 1][j - 1] + dp[i - 1][j]
                else:
                    dp[i][j] = dp[i - 1][j]

        return dp[-1][-1]
